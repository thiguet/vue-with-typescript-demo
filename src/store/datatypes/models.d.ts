
export enum Measures {
    'UN',
    'L',
    'KG',
    'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}
