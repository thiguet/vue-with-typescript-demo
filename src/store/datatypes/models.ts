export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}

export interface User {
    name: string;
    email: string;
}

export interface LoginSubmit {
    name: string;
    pass: string;
}

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
}
