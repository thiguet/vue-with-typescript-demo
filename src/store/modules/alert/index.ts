import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes as MTypes } from './mutations';

export const MutationTypes = MTypes;
export const AlertMutations = Mutations;

export class AlertState {
  message!: string;

  display!: boolean;
}

export default new Module({
    state: AlertState,
    mutations: AlertMutations,
});
