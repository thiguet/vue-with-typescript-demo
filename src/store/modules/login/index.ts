import { Module } from 'vuex-smart-module';
import LoginMutations, { MutationTypes as MTypes } from './mutations';
import LoginActions, { ActionTypes as ATypes } from './actions';
import State from './state';

export type LoginState = State;
export const Mutations = LoginMutations;
export const MutationTypes = MTypes;
export const Actions = LoginActions;
export const ActionTypes = ATypes;

export default new Module({
    state: State,
    mutations: LoginMutations,
    actions: LoginActions,
});
