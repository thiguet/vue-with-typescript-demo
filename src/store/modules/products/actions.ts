import { Actions as VActions, Getters } from 'vuex-smart-module';
import { Product } from '@/store/datatypes/models.d';
import State from './state';
import Mutations, { MutationTypes } from './mutations';

export enum ActionTypes {
    newProduct = 'newProduct',
}

export default class Actions extends VActions<
  State,
  {} & Getters<State>,
  Mutations,
  Actions> {
  [ActionTypes.newProduct](payload: Product) {
    this.commit(MutationTypes.addProduct, payload);
  }
}
