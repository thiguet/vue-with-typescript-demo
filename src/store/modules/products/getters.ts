import { Getters as VGetters } from 'vuex-smart-module';
import State from './state';

export enum GettersTypes {
    tableRows = 'tableRows',
}
export default class Getters extends VGetters<State> {
    get [GettersTypes.tableRows]() {
        return this.state.products.map((p) => ({
            icon: p.image,
            name: p.name,
        }));
    }
}
