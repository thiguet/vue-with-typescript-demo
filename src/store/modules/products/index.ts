import { Module } from 'vuex-smart-module';
import State from './state';
import ProductsMutations, { MutationTypes as MTypes } from './mutations';
import ProductsActions, { ActionTypes as ATypes } from './actions';

export type ProductsState = State;
export const MutationTypes = MTypes;
export const ActionTypes = ATypes;

export default new Module({
  state: State,
  mutations: ProductsMutations,
  actions: ProductsActions,
});
