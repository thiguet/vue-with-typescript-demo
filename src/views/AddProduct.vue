<template>
    <div class="add-product">
        <div class="img-wrapper">
            <div class="img-container">
                <span>Img:</span>
                <img id="img" />
            </div>
            <Button
                id="img-btn"
                name="img-btn"
                label="Escolha uma imagem"
                :onclick="() => {}"
            />
        </div>
        <div class="form">
            <form action="#">
                <InputWrapper label="Nome" legend="Product Name">
                    <TextInput
                        id="name"
                        :value="ProductName"
                        :setValue="ProductName"
                    />
                </InputWrapper>
                <InputWrapper label="Quantidade" legend="Product Quantity">
                    <TextInput
                        id="qtd"
                        type="number"
                        :value="ProductQtd"
                        :setValue="ProductQtd"
                    />
                </InputWrapper>
                <InputWrapper label="Qtd (Min)" legend="Product Min Quantity">
                    <TextInput
                        id="minQtd"
                        type="number"
                        :value="ProductMinQtd"
                        :setValue="ProductMinQtd"
                    />
                </InputWrapper>
                <InputWrapper
                    label="Product Measure"
                    legend="Product Measure"
                    for="measure"
                >
                    <select
                        name="measure"
                        id="measure"
                        @change="Measure($event.target.value)"
                    >
                        <option selected disabled>Escolha uma opção</option>
                        <option
                            v-for="(measure, index) in measures"
                            :key="index"
                        >
                            {{ measure }}
                        </option>
                    </select>
                </InputWrapper>
                <Button
                    id="add-product"
                    name="add-product"
                    label="Enviar"
                    :onclick="() => {}"
                />
            </form>
        </div>
    </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

import InputWrapper from '@/components/InputWrapper.vue';
import TextInput from '@/components/TextInput.vue';
import Button from '@/components/Button.vue';

import { namespace } from 'vuex-class';

import { AddProductView } from '@/views/models.d';

import { Product, Measures, VuexAppModules } from '@/store/datatypes/models';

import Mutations, { MutationTypes } from '@/store/modules/products/mutations';

const products = namespace(VuexAppModules.products);

@Component({
    name: 'AddProduct',
    components: {
        InputWrapper,
        TextInput,
        Button,
    },
    data: () => ({
        measures: Object.values(Measures),
    }),
})
export default class AddProduct extends Vue implements AddProductView {
    @products.State
    private selectedProduct!: Product;

    @products.Mutation
    setProduct!: Mutations[MutationTypes.setProduct];

    get ProductName() {
        return ''; // this.selectedProduct.name;
    }

    set ProductName(val: string) {
        this.setProduct({
            ...this.selectedProduct,
            name: val,
        });
    }

    get Measure() {
        return this.selectedProduct.measure;
    }

    set Measure(val: keyof typeof Measures) {
        this.setProduct({
            ...this.selectedProduct,
            measure: val,
        });
    }

    public newProduct() {
        console.log('a');
    }

    public setMeasure() {
        this.setProduct({
            ...this.selectedProduct,
            // measure: value,
        });
    }
}
</script>

<style scoped>
select > option,
select#measure {
    font-size: 16px;
}

.img-container {
    display: flex;
    padding: 50px 30vw;
    margin-bottom: 5px;
    border: 1px solid var(--border-color);
}

.form {
    display: flex;
    flex-flow: row;
    width: 100%;
    justify-content: space-around;
    align-self: center;
    align-content: center;
    align-items: center;
}
.form > form {
    display: relative;
    width: 100%;
}

form > fieldset {
    width: 70vw;
}

.add-product {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    align-items: center;
}

.img-wrapper {
    flex-direction: column;
    width: 100%;
}

select {
    width: 100%;
    border: none;
    border-bottom: 1px solid var(--border-color);
    font-family: inherit;
    font-size: inherit;
}

fieldset {
    padding: 0;
    display: flex;
    width: 100%;
    border: none;
}

label {
    width: 100%;
}
</style>
