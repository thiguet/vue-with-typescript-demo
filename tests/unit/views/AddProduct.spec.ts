import { shallowMount, createLocalVue, Wrapper } from '@vue/test-utils';
import AddProduct from '@/views/AddProduct.vue';
import Vuex from 'vuex';
import faker from 'faker';
import Store from '@/store';
import { Measures } from '@/store/datatypes/models';
import { ProductsState } from '@/store/modules/products';
import {
    ButtonProps,
    NumberInputProps,
    TextInputProps,
} from '../components/models';
import { keys } from 'ts-transformer-keys';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('AddProduct.vue', () => {
    let store: typeof Store;
    let state: ProductsState;

    const build = () => {
        const wrapper = shallowMount(AddProduct, {
            localVue,
            store,
        });

        return {
            wrapper,
            imgContainer: () => wrapper.find('.img-container'),
            imgBtn: () => wrapper.find('#img-btn'),
            img: () => wrapper.find('#img'),
            name: () => wrapper.find('#name'),
            measure: () => wrapper.find('#measure'),
            qtd: () => wrapper.find('#qtd'),
            minQtd: () => wrapper.find('#minQtd'),
            submit: () => wrapper.find('#add-product'),
        };
    };

    const getRandomMeasure = () =>
        faker.random.arrayElement(Object.values(Measures)) as Measures;

    beforeEach(() => {
        state = {
            selectedProduct: {
                name: faker.random.word(),
                measure: getRandomMeasure(),
                qtd: faker.random.number(),
                minQtd: faker.random.number(),
            },
            products: [],
        };
    });

    it('renders the component', () => {
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main components', () => {
        const {
            imgBtn,
            name,
            qtd,
            minQtd,
            measure,
            submit,
            imgContainer,
        } = build();

        expect(name().exists()).toBe(true);
        expect(qtd().exists()).toBe(true);
        expect(minQtd().exists()).toBe(true);
        expect(measure().exists()).toBe(true);
        expect(submit().exists()).toBe(true);
        expect(imgBtn().exists()).toBe(true);
        expect(imgContainer().exists()).toBe(true);
    });

    it('check types for inputs', () => {
        const { name, qtd, minQtd, submit, imgBtn } = build();

        // const TextKeys = keys<TextInputProps>();
        // const NumberKeys = keys<NumberInputProps>();
        // const ButtonKeys = keys<ButtonProps>();

        // const getPropsKeys = (comp: Wrapper<Vue>) => Object.keys(comp.props());

        // expect(getPropsKeys(name())).toBe(TextKeys);
        // expect(getPropsKeys(qtd())).toBe(NumberKeys);
        // expect(getPropsKeys(minQtd())).toBe(NumberKeys);
        // expect(getPropsKeys(imgBtn())).toBe(ButtonKeys);
        // expect(getPropsKeys(submit())).toBe(ButtonKeys);
    });
});
