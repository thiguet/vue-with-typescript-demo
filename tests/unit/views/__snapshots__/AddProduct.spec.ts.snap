// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct.vue renders the component 1`] = `
<div class="add-product">
  <div class="img-wrapper">
    <div class="img-container"><span>Img:</span> <img id="img"></div>
    <button-stub label="Escolha uma imagem" onclick="function () {}" id="img-btn" name="img-btn"></button-stub>
  </div>
  <div class="form">
    <form action="#">
      <inputwrapper-stub label="Nome" legend="Product Name">
        <textinput-stub id="name" type="text" value="" setvalue=""></textinput-stub>
      </inputwrapper-stub>
      <inputwrapper-stub label="Quantidade" legend="Product Quantity">
        <textinput-stub id="qtd" type="number"></textinput-stub>
      </inputwrapper-stub>
      <inputwrapper-stub label="Qtd (Min)" legend="Product Min Quantity">
        <textinput-stub id="minQtd" type="number"></textinput-stub>
      </inputwrapper-stub>
      <inputwrapper-stub label="Product Measure" legend="Product Measure" for="measure"><select name="measure" id="measure">
          <option selected="selected" disabled="disabled">Escolha uma opção</option>
          <option>
            UN
          </option>
          <option>
            L
          </option>
          <option>
            KG
          </option>
          <option>
            G
          </option>
        </select></inputwrapper-stub>
      <button-stub label="Enviar" onclick="function () {}" id="add-product" name="add-product"></button-stub>
    </form>
  </div>
</div>
`;
